atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreIntegrationTest.java

```diff
@@ -0,0 +1,247 @@
+        assertThat(store.getUpperLimit()).isEqualTo(Optional.of(TIMESTAMP_1));
+    }
+
+    @Test
+    public void throwsIfStoringBoundWithExpectationWhenThereIsNone() {
+        assertThatThrownBy(() -> store.storeTimestampBound(Optional.of(TIMESTAMP_1), TIMESTAMP_2))
+                .isInstanceOf(ConcurrentModificationException.class);
+    }
+
+    @Test
+    public void throwsIfStoringBoundWithIncorrectExpectation() {
+        store.storeTimestampBound(Optional.empty(), TIMESTAMP_1);
+        assertThatThrownBy(() -> store.storeTimestampBound(Optional.of(TIMESTAMP_2), TIMESTAMP_3))
+                .isInstanceOf(ConcurrentModificationException.class);
+    }
+
+    @Test
+    public void throwsIfReadingInvalidatedTimestampBound() {
+        store.backupExistingTimestamp(TIMESTAMP_1);
+        assertThatThrownBy(store::getUpperLimit).isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
+    public void throwsIfStoringToABackedUpStore() {
+        store.storeTimestampBound(Optional.empty(), TIMESTAMP_1);
+        assertThat(store.backupExistingTimestamp(TIMESTAMP_3)).isEqualTo(TIMESTAMP_1);
+        assertThatThrownBy(() -> store.storeTimestampBound(Optional.of(TIMESTAMP_1), TIMESTAMP_2))
+                .isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
+    public void throwsOnBackupOrRestoreWithTwoReadableBounds() {
+        store.backupExistingTimestamp(TIMESTAMP_1);
+        byte[] rowAndColumnNameBytes = PtBytes.toBytes(CassandraTimestampUtils.ROW_AND_COLUMN_NAME);
+        kv.put(AtlasDbConstants.TIMESTAMP_TABLE,
+                ImmutableMap.of(Cell.create(rowAndColumnNameBytes, rowAndColumnNameBytes),
+                        PtBytes.toBytes(TIMESTAMP_2)),
+                Long.MAX_VALUE - 1);
+        assertThatThrownBy(() -> store.backupExistingTimestamp(TIMESTAMP_2))
+                .isInstanceOf(IllegalStateException.class);
+        assertThatThrownBy(store::restoreFromBackup).isInstanceOf(IllegalStateException.class);
+    }
+
+    @Test
+    public void backsUpStoredBound() {
+        store.storeTimestampBound(Optional.empty(), TIMESTAMP_1);
+        store.backupExistingTimestamp(TIMESTAMP_2);
+        store.restoreFromBackup();
+        assertThat(store.getUpperLimit()).isEqualTo(Optional.of(TIMESTAMP_1));
+    }
+
+    @Test
+    public void backsUpDefaultValueIfNoBoundExists() {
+        store.backupExistingTimestamp(TIMESTAMP_1);
+        store.restoreFromBackup();
+        assertThat(store.getUpperLimit()).isEqualTo(Optional.of(TIMESTAMP_1));
+    }
+
+    @Test
+    public void firstWriteWinsInPresenceOfMultipleBackups() {
+        assertThat(store.backupExistingTimestamp(TIMESTAMP_1)).isEqualTo(TIMESTAMP_1);
+        assertThat(store.backupExistingTimestamp(TIMESTAMP_2)).isEqualTo(TIMESTAMP_1);
+        store.restoreFromBackup();
+        assertThat(store.getUpperLimit()).isEqualTo(Optional.of(TIMESTAMP_1));
+    }
+
+    @Test
+    public void canRestoreWithoutBackupWithoutData() {
```

Comment:
Ah I missed this when changing the behaviour, thanks! Now `throwsIfRestoringFromUnavailableBackup`.